### PR TITLE
Fix recovery middleware to render gitea style page.

### PR DESCRIPTION
**Pull request by lunny**
The recent change to recovery middleware will always output trace log content to web browser when panic. It's a regression.

This PR introduced some new middlewares to chi and will fix that and take them back.

- [x] Introduce new render middleware from https://github.com/unrolled/render.
- [x] Introduce new locale middleware which could get the current language.
- [x] Session middleware has been copied and modified to https://gitea.com/go-chi/session.
- [x] Login user panic page should know user information.
- [x] code.gitea.io/gitea/modules/auth/sso has been refactored so that both macaron and chi could use that.
- [x] Improve the Recovery middleware, now it will try to render a friendly UI at first, if panic again, fallback to original method

### CODE REVIEW

File: modules/middlewares/cookie.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		Value:    value,
+		HttpOnly: true,
+		Path:     setting.SessionConfig.CookiePath,
+		Domain:   setting.SessionConfig.Domain,
+		MaxAge:   maxAge,
+		Secure:   setting.SessionConfig.Secure,
+	}
+}
+
+// SetCookie set the cookies
```
Comment (6543) [inline modules/middlewares/cookie.go:28]: this is from macaron so we have to let it this way untill we can remove macaron.

But @lunny can you add a TODO note to clean up this messy func after switching?

File: modules/middlewares/redis.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
+// Copyright 2013 Beego Authors
+// Copyright 2014 The Macaron Authors
+// Copyright 2020 The Gitea Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"): you may
+// not use this file except in compliance with the License. You may obtain
```
Comment (6543) [inline modules/middlewares/redis.go:6]: looks like a copy of https://gitea.com/go-chi/session/src/branch/master/redis/redis.go

do we have to have it twice?

File: modules/templates/static.go [hunk 3]
```diff
@@ -47,6 +49,29 @@ func (templates templateFileSystem) Get(name string) (io.Reader, error) {
 	return nil, fmt.Errorf("file '%s' not found", name)
 }
 
+func GetAsset(name string) ([]byte, error) {
```
Comment (6543) [inline modules/templates/static.go:52]: lint err

File: modules/middlewares/redis.go (hunk 2 above)
Comment (lunny) [reply modules/middlewares/redis.go:6]: Copied from modules/session/redis.go and should remove that one until macaron removed. I will add comment here.

File: modules/templates/static.go (hunk 3 above)
Comment (lunny) [reply modules/templates/static.go:52]: I have changed this one. But it seems lint check on this file is ignored because of some reason.

**Comment by lunny** [comment]
@6543 All done.

**Review by 6543 — APPROVED** [review]

File: modules/templates/base.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		}
+		log.Warn("Unable to check if templates dir %s is a directory. Error: %v", dir, err)
+		return tmpls
+	}
+	if !f.IsDir() {
+		log.Warn("Templates dir %s is a not directory.", dir)
+		return tmpls
+	}
+
+	files, err := com.StatDir(dir)
```
Comment (techknowlogick) [inline modules/templates/base.go]: Please use `code.gitea.io/gitea/modules/util` path functions instead of `com`

Comment (lunny) [reply modules/templates/base.go]: @techknowlogick Can't find that function on `code.gitea.io/gitea/modules/util`, maybe you mean `util.IsDir`?

Comment (6543) [reply modules/templates/base.go]: If I remember correctly this was done because of better error handling ...

Comment (6543) [reply modules/templates/base.go]: -> https://github.com/go-gitea/gitea/pull/13186

Comment (lunny) [reply modules/templates/base.go]: But no `StatDir` there.

Comment (6543) [reply modules/templates/base.go]: from Discord: "Could we add a function to the util package? I think we should stop relying on com"

**Review by 6543 — APPROVED** [review]

**Comment by lunny** [comment]
@techknowlogick Could you review this again?

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
:tada:

**Pull request merged**